data/restli-csharp/datagenerator/src/test/GeneratedDataTemplate/com/linkedin/restli/datagenerator/integration/NestedReallySimpleRecord.cs

```diff
@@ -0,0 +1,199 @@
+using System.Collections.Generic;
+using System;
+using restlicsharpdata.restlidata;
+
+// DO NOT EDIT - THIS FILE IS GENERATED BY restli-csharp
+// Generated from com/linkedin/restli/datagenerator/integration/NestedReallySimpleRecord.pdsc
+
+namespace com.linkedin.restli.datagenerator.integration
+{
+  
+  public class NestedReallySimpleRecord : RecordTemplate
+  {
+
+    // optional
+    public SimpleRecord simpleRecordValue { get; }
+
+
+    // optional, has default value
+    public IReadOnlyList<SimpleRecord> collectionWithDefault { get; }
+    public bool hasCollectionWithDefault { get; }
+
+    // required
+    public IReadOnlyList<IReadOnlyList<IReadOnlyList<SimpleRecord>>> triple { get; }
+
+
+    // optional
+    public IReadOnlyDictionary<string, IReadOnlyDictionary<string, SimpleRecordProjection>> dict { get; }
+
+
+    // optional
+    public IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<SimpleRecord>>> mixed { get; }
+
+
+    public NestedReallySimpleRecord(Dictionary<string, object> data)
+    {
+      object value;
+      // Retrieve data for simpleRecordValue
+      if (data.TryGetValue("simpleRecordValue", out value))
+      {
+        simpleRecordValue = new SimpleRecord((Dictionary<string, object>)value);
+
+      }
+
+      // Retrieve data for collectionWithDefault
+      if (data.TryGetValue("collectionWithDefault", out value))
+      {
+        List<Dictionary<string, object>> data0 = (List<Dictionary<string, object>>)value;
+        List<SimpleRecord> result0 = new List<SimpleRecord>();
+        foreach (Dictionary<string, object> data1 in data0)
+        {
+          SimpleRecord result1;
+          result1 = new SimpleRecord((Dictionary<string, object>)data1);
+          result0.Add(result1);
+        }
+        collectionWithDefault = result0;
+        hasCollectionWithDefault = true;
+      }
+      else
+      {
+        collectionWithDefault = new List<SimpleRecord>();
+        hasCollectionWithDefault = false;
+      }
+      // Retrieve data for triple
+      if (data.TryGetValue("triple", out value))
+      {
+        List<List<List<Dictionary<string, object>>>> data0 = (List<List<List<Dictionary<string, object>>>>)value;
+        List<IReadOnlyList<IReadOnlyList<SimpleRecord>>> result0 = new List<IReadOnlyList<IReadOnlyList<SimpleRecord>>>();
+        foreach (List<List<Dictionary<string, object>>> data1 in data0)
+        {
+          List<IReadOnlyList<SimpleRecord>> result1 = new List<IReadOnlyList<SimpleRecord>>();
+          foreach (List<Dictionary<string, object>> data2 in data1)
+          {
+            List<SimpleRecord> result2 = new List<SimpleRecord>();
+            foreach (Dictionary<string, object> data3 in data2)
+            {
+              SimpleRecord result3;
+              result3 = new SimpleRecord((Dictionary<string, object>)data3);
+              result2.Add(result3);
+            }
+            result1.Add(result2);
```

Comment:
Does C# do  automatic casting here when we add a non read-only List to result1 (which is of List<IReadOnlyList>)?